### ISSUE TITLE
[BUG]: Stop all operations to resolve the SERVICE IS BUSY error

### ISSUE THREAD

**Issue by jeremielodi** [ksenia312/nearby_service#6]
I got this error bellow while trying to discover peers.
Then phone was able to discover before but now I got an error.

<img width="1538" alt="Screenshot 2024-05-02 at 13 38 39" src="https://github.com/ksenia312/nearby_service/assets/25838121/b7c504de-e7b4-4a0f-a7df-281ab47b0ed1">

**Comment by jeremielodi** [comment]
@ksenia312 , when I try to use this following function ```context.read<AppService>().stopListeningAll()``` I face the same issue. The ```NearbyService``` is still ```busy```

**Comment by ksenia312** [comment]
Yes, you're right
The problem here is that you have a pending request to another device.
It does not depend on your active subscriptions, and stopping listening will not help

It will be solved with functionality to cancel the request, and service will become free 

I'm going to introduce it soon. Now you can catch this error and wait some time while the request is complete, or implement retry logic for this 

Sorry for that🙌

**Comment by jeremielodi** [comment]
Thanks @ksenia312, looking forward to have that option implemented !

**Comment by jeremielodi** [comment]
Think, you are talking about this ```operation``` bellow :


```js

val manager = getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
val channel = manager.initialize(this, mainLooper, null)

manager.cancelConnect(channel, object : WifiP2pManager.ActionListener {
    override fun onSuccess() {
        // Connection request canceled successfully
    }

    override fun onFailure(reason: Int) {
        // Handle failure to cancel connection request
    }
})

```

**Comment by ksenia312** [comment]
> Think, you are talking about this `operation` bellow :
> 
> ```js
> val manager = getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
> val channel = manager.initialize(this, mainLooper, null)
> 
> manager.cancelConnect(channel, object : WifiP2pManager.ActionListener {
>     override fun onSuccess() {
>         // Connection request canceled successfully
>     }
> 
>     override fun onFailure(reason: Int) {
>         // Handle failure to cancel connection request
>     }
> })
> ```

Yes, definitely

**Comment by jeremielodi** [comment]
Hello @ksenia312,  did you fix this issue, we are looking forward is option, please!!!

**Comment by ksenia312** [comment]
> Hello @ksenia312, did you fix this issue, it's extremely important to help easy sharing between two flutter apps, please!!!

I plan to do it around May 15-20. Sorry, very busy now, I will definitely do a plugin update, but there's no way I'll be able to do it sooner than that

For now, you can use the options described above

**Comment by ksenia312** [comment]
Hello @jeremielodi 

Update is available on https://pub.dev/packages/nearby_service

In android implementation there is a new method `cancelLastConnectionProcess()`. Use it as `_nearbyService.android?.cancelLastConnectionProcess()`.

Check if everything is OK and close the issue if so!

**Comment by jeremielodi** [comment]
Hello @ksenia312 ,
I faced this bug while calling the function ```cancelLastConnectionProcess()```
<img width="1246" alt="Screenshot 2024-05-20 at 00 39 13" src="https://github.com/ksenia312/nearby_service/assets/25838121/3eed114d-6fe7-40d3-b2f1-515687979336">

**Comment by jeremielodi** [comment]
<img width="1520" alt="Screenshot 2024-05-20 at 08 57 38" src="https://github.com/ksenia312/nearby_service/assets/25838121/4a922ce2-54a6-4f17-a1c3-15c14c4e464c">

**Comment by ksenia312** [comment]
Hello, @jeremielodi 
Could you please provide some details of your code? I didn't get this error during cancellation, I need more information to fix it

**Comment by jeremielodi** [comment]
Okay, I added this function in the app_service.dart, 
<img width="605" alt="Screenshot 2024-05-20 at 09 31 37" src="https://github.com/ksenia312/nearby_service/assets/25838121/5e967078-d8fd-4521-97c1-a2d0d164e8ea">

I call the function before initialising the service or after. I get the same error.
<img width="738" alt="Screenshot 2024-05-20 at 09 34 00" src="https://github.com/ksenia312/nearby_service/assets/25838121/5ce3a35f-d738-47ef-8e9e-ef6f45ad4fe5">


What is curious, even when I uninstall the app and reinstall it, I face the same issue.
I wander if you can have an idea from those informations

**Comment by ksenia312** [comment]
You should call `cancelLastConnectionProcess()` when you have any peer with status `CONNECTING`
Otherwise, of course, it can lead to unexpected behavior because it doesn't really have any process to cancel

You can add this check:
```dart
 Future<bool> hasRunningJobs() async {
    try {
      final result = await _nearbyService.getPeers();
      // if one of devices is connecting, service is busy (android only)
      if (result.any(
        (element) => element.status == NearbyDeviceStatus.connecting,
      )) {
        if (kDebugMode) {
          print('Service has already running jobs');
        }
        return true;
      }
      return false;
    } catch (e, s) {
      _log(e, s);
      return false;
    }
  }
```
Or just catch and ignore this error, because predictably `cancelLastConnectionProcess()` will fail if we do not have any process to cancel

**Comment by jeremielodi** [comment]
the function ```hasRunningJobs``` return ```false``` but the servis is still busy. Curious

**Comment by ksenia312** [comment]
You shouldn't call cancelLastConnectionProcess() if hasRunningJobs() is false

**Comment by jeremielodi** [comment]
Yes, you are right, waht is complicated to know is why the service status is busy at the time there is no running job.

**Comment by ksenia312** [comment]
I agree that it is not really clear and I don't know why it is busy haha

I just provide the error from platform to flutter 
That's reason code that platform returns, I actually have no idea what we can do with it

Seems like platform returns this code as indicator that there is no requests to be cancelled. BUSY is very common error for all situations when we use wifiP2pManager in a wrong way maybe

![image](https://github.com/ksenia312/nearby_service/assets/71008947/c2e77c9b-c812-4a1a-99c5-32119b8acac3)
